File: C#/第二套/11900-1060307/Services/PokerService.cs
namespace PokerGame.Services;

using PokerGame.Models;

public interface IPokerService
{
    RoundResult PlayRound(int roundIndex, double playerRandom, double bankerRandom);
}

public class PokerService : IPokerService
{
    public RoundResult PlayRound(int roundIndex, double playerRandom, double bankerRandom)
    {
        int playerCardNum = (int)(playerRandom * 52);
        int bankerCardNum = (int)(bankerRandom * 52);

        // Requirement: If same card is picked, player picks the next available random number?
        // PDF p19: "但須注意同一張牌不可發出兩次或兩次以上，遇取得之牌張已發出時，則捨棄該牌張，重新另取一張。"
        // Since we are given a fixed sequence of random numbers from a file, if bankerCardNum == playerCardNum, 
        // we should conceptually take the *next* random number from the file.
        // This implies the PlayRound logic needs to be handled by the Presenter who reads the file.

        var playerCard = new Card(playerCardNum);
        var bankerCard = new Card(bankerCardNum);

        string result = "平手";
        if (playerCard.PowerValue > bankerCard.PowerValue)
        {
            result = "玩家贏";
        }
        else if (playerCard.PowerValue < bankerCard.PowerValue)
        {
            result = "莊家贏";
        }

        return new RoundResult(roundIndex, playerCard, bankerCard, result);
    }
}

File: VB/第二套/11900-1060307/Services/PokerService.vb
Namespace Models
    Public Class RoundResult
        Public Property RoundIndex As Integer
        Public Property PlayerCard As Card
        Public Property BankerCard As Card
        Public Property Result As String

        Public Sub New(idx As Integer, p As Card, b As Card, res As String)
            Me.RoundIndex = idx
            Me.PlayerCard = p
            Me.BankerCard = b
            Me.Result = res
        End Sub
    End Class
End Namespace

Namespace Services
    Public Interface IPokerService
        Function PlayRound(roundIndex As Integer, playerRandom As Double, bankerRandom As Double) As Models.RoundResult
    End Interface

    Public Class PokerService
        Implements IPokerService

        Public Function PlayRound(roundIndex As Integer, playerRandom As Double, bankerRandom As Double) As Models.RoundResult Implements IPokerService.PlayRound
            Dim playerCardNum As Integer = Int(playerRandom * 52)
            Dim bankerCardNum As Integer = Int(bankerRandom * 52)

            Dim pCard As New Models.Card(playerCardNum)
            Dim bCard As New Models.Card(bankerCardNum)

            Dim resStr As String = "平手"
            If pCard.PowerValue > bCard.PowerValue Then
                resStr = "玩家贏"
            ElseIf pCard.PowerValue < bCard.PowerValue Then
                resStr = "莊家贏"
            End If

            Return New Models.RoundResult(roundIndex, pCard, bCard, resStr)
        End Function
    End Class
End Namespace
